File: java/LongestSubstringWithoutRepeatingCharacters.java
/*
 * Filename: LongestSubstringWithoutRepeatingCharacters.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Tue 03 Mar 2015 12:36:06 PM EST
 */

public class Solution {
	public static int lengthOfLongestSubstring(String s) {
		HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
		Integer c;

		int start = 0;
		int end = 0;
		int length = 0;

		for (; start != s.length(); start++) {
			c = new Integer(s.charAt(start));
			if (map.containsKey(c)) {
				if (length < start - end)
					length = start - end;
				int temp = map.get(c) + 1;
				end = temp > end ? temp : end;
				map.remove(c);
			}
			map.put(c, start);
		}

		if (length < start - end)
			length = start - end;

		return length;
	}
}

File: cpp/longest_substring_without_repeating_characters.cc
/*
 * Filename: longest_substring_without_repeating_characters.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 19 Aug 2015 07:37:16 PM EDT
 */

#include <climits>
#include <iostream>
#include <map>
#include <string>

using namespace std;

class Solution {
	public:
		int lengthOfLongestSubstring(string s) {
			map<char, int> mapping;
			int res = 0;
			int j = 0;

			for (int i = 0; i < s.length(); i++)
			{
				if (mapping.find(s[i]) != mapping.end())
				{
					res = max(res, i-j);
					j = max(mapping[s[i]] + 1, j);
				}

				mapping[s[i]] = i;
			}

			return max(res, (int) s.length() - j);
		}
};
